core/trino-main/src/test/java/io/trino/sql/planner/iterative/rule/TestRewriteCaseExpressionPredicate.java

```diff
@@ -0,0 +1,200 @@
+
+        assertRewrittenAndOptimizedExpression(
+                "25 < (case col1 when 1 then 10 when 2 then 20 else 30 end)",
+                "(25 < 10 AND col1 IS NOT NULL AND col1 = 1) OR (25 < 20 AND col1 IS NOT NULL AND col1 = 2) OR (25 < 30 AND (col1 IS NULL OR (NOT(col1 = 1) AND NOT(col1 = 2))))");
+    }
+
+    @Test
+    public void testRewriterWhenMoreThanOneConditionMatches()
+    {
+        assertRewrittenAndOptimizedExpression(
+                "(case col1 when 1 then 'case' when 2 then 'case' else 'default' end) = 'case'",
+                "('case' = 'case' AND col1 IS NOT NULL AND col1 = 1) OR ('case' = 'case' AND col1 IS NOT NULL AND col1 = 2) OR ('default' = 'case' AND (col1 IS NULL OR (NOT(col1 = 1) AND NOT(col1 = 2))))");
+
+        assertRewrittenAndOptimizedExpression(
+                "(case col1 when 1 then 'defaultAndCase1' when 2 then 'case2' else 'defaultAndCase1' end) = 'defaultAndCase1'",
+                "('defaultAndCase1' = 'defaultAndCase1' AND col1 IS NOT NULL AND col1 = 1) OR ('case2' = 'defaultAndCase1' AND col1 IS NOT NULL AND col1 = 2) OR ('defaultAndCase1' = 'defaultAndCase1' AND (col1 IS NULL OR (NOT(col1 = 1) AND NOT(col1 = 2))))");
+
+        assertRewrittenAndOptimizedExpression(
+                "(case col3 when 'data1' then 'case1' when 'data2' then 'case2' else col3 end) = 'case1'",
+                "('case1' = 'case1' AND col3 IS NOT NULL AND col3 = 'data1') OR ('case2' = 'case1' AND col3 IS NOT NULL AND col3 = 'data2') OR (col3 = 'case1' AND (col3 IS NULL OR (NOT(col3 = 'data1') AND NOT(col3 = 'data2'))))");
+    }
+
+    @Test
+    public void testRewriterOnCaseExpressionWithoutElseClause()
+    {
+        assertRewrittenAndOptimizedExpression(
+                "(case col1 when 1 then 'case1' when 2 then 'case2' end) = 'case1'",
+                "('case1' = 'case1' AND col1 IS NOT NULL AND col1 = 1) OR ('case2' = 'case1' AND col1 IS NOT NULL AND col1 = 2) OR (CAST(null as VARCHAR(5)) = 'case1' AND (col1 IS NULL OR (NOT(col1 = 1) AND NOT(col1 = 2))))");
+
+        assertRewrittenAndOptimizedExpression(
+                "(case col1 when 1 then 'case1' when 2 then 'case2' end) = 'case3'",
+                "('case1' = 'case3' AND col1 IS NOT NULL AND col1 = 1) OR ('case2' = 'case3' AND col1 IS NOT NULL AND col1 = 2) OR (CAST(null as VARCHAR(5)) = 'case3' AND (col1 IS NULL OR (NOT(col1 = 1) AND NOT(col1 = 2))))");
+
+        assertRewrittenAndOptimizedExpression(
+                "(case col1 when 1 then 'case1' when 2 then 'case2' end) = 'case2'",
+                "('case1' = 'case2' AND col1 IS NOT NULL AND col1 = 1) OR ('case2' = 'case2' AND col1 IS NOT NULL AND col1 = 2) OR (CAST(null as VARCHAR(5)) = 'case2' AND (col1 IS NULL OR (NOT(col1 = 1) AND NOT(col1 = 2))))");
+    }
+
+    @Test
+    public void testRewriterOnCaseExpressionWithCastFunction()
```

Comment:
> @maswin
> 
> ```
> (result1 = value AND expression IS NOT NULL AND expression=constant1) OR
> (result2 = value AND expression IS NOT NULL AND expression=constant2) OR
> (result3 = value AND expression IS NOT NULL AND expression=constant3) OR
> (elseResult = value AND ((expression IS NULL) OR (!(expression=constant1) AND !(expression=constant2) AND !(expression=constant3)))
> ```
> 
> Adding `expression IS NOT NULL` in the branches 1, 2 and 3 is redundant. They will fail anyway for `expression` being `null` because `expression=constant` returns null. The only thing we really need is the additional condition in the last branch.

@kasiafi : The idea of this optimization is most of the 'result=value' expression will evaluate to false during the expression simplification and most of the time only one branch will be present in the end. 
The problem with adding IS NULL check only in last branch is, the last branch might be discarded by ExpressionInterpreter during simplification and NULL check will also get discarded. So we need the NULL check in all branches. 

For example, if we add NULL check only to last branch

```
(CASE
  WHEN col = 1 THEN 'a'
  WHEN col = 2 THEN 'b'
  ELSE 'c'
END) = 'a'
```
will be converted to
```
('a'='a' AND col = 1) OR
('b'='a' AND col = 2) OR
('c'='a' AND ((col IS NULL) OR ((col != 1) AND (col != 2)))
```
after simplification will become
`col = 1`

which will result in NULL=1 and return NULL, but we need 'FALSE' to be returned. By adding IS NOT NULL check, the simplification would be
`col IS NOT NULL and col=1`
which will make sure FALSE is returned instead of NULL.